File: mobile/src/main/java/treehou/se/habit/ui/control/cells/config/CellIncDecConfigFragment.java
package treehou.se.habit.ui.control.cells.config;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;

import com.trello.rxlifecycle.components.support.RxFragment;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import io.realm.Realm;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import se.treehou.ng.ohcommunicator.connector.models.OHItem;
import se.treehou.ng.ohcommunicator.connector.models.OHServer;
import se.treehou.ng.ohcommunicator.services.IServerHandler;
import treehou.se.habit.R;
import treehou.se.habit.core.db.model.ItemDB;
import treehou.se.habit.core.db.model.ServerDB;
import treehou.se.habit.core.db.model.controller.CellDB;
import treehou.se.habit.core.db.model.controller.IncDecCellDB;
import treehou.se.habit.util.ConnectionFactory;
import treehou.se.habit.util.Constants;
import treehou.se.habit.util.Util;
import treehou.se.habit.ui.util.IconPickerActivity;

public class CellIncDecConfigFragment extends RxFragment {

    private static final String TAG = "CellIncDecConfigFragment";

    private static String ARG_CELL_ID = "ARG_CELL_ID";
    private static int REQUEST_ICON = 183;

    @BindView(R.id.spr_items) Spinner sprItems;
    @BindView(R.id.txtMax) EditText txtMax;
    @BindView(R.id.txtMin) EditText txtMin;
    @BindView(R.id.txtValue) EditText txtValue;
    @BindView(R.id.btn_set_icon) ImageButton btnSetIcon;

    @Inject ConnectionFactory connectionFactory;
    Realm realm;

    private ArrayAdapter<OHItem> itemAdapter;
    private ArrayList<OHItem> items = new ArrayList<>();

    private IncDecCellDB incDecCell;
    private CellDB cell;
    private OHItem item;
    private Unbinder unbinder;

    public static CellIncDecConfigFragment newInstance(CellDB cell) {
        CellIncDecConfigFragment fragment = new CellIncDecConfigFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_CELL_ID, cell.getId());
        fragment.setArguments(args);
        return fragment;
    }

    public CellIncDecConfigFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Util.getApplicationComponent(this).inject(this);
        realm = Realm.getDefaultInstance();

        if (getArguments() != null) {
            long id = getArguments().getLong(ARG_CELL_ID);
            cell = CellDB.load(realm, id);
            incDecCell = cell.getCellIncDec();

            if (incDecCell == null) {
                realm.executeTransaction(realm -> {
                    incDecCell = new IncDecCellDB();
                    incDecCell = realm.copyToRealm(incDecCell);
                    cell.setCellIncDec(incDecCell);
                    realm.copyToRealmOrUpdate(cell);
                });
            }

            ItemDB itemDB = incDecCell.getItem();
            if(itemDB != null){
                item = itemDB.toGeneric();
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.inc_dec_controller_action, container, false);
        unbinder = ButterKnife.bind(this, rootView);

        txtMax.setText("" + incDecCell.getMax());
        txtMin.setText("" + incDecCell.getMin());
        txtValue.setText("" + incDecCell.getValue());

        sprItems.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                realm.beginTransaction();
                OHItem item = items.get(position);
                if(item != null) {
                    ItemDB itemDB = ItemDB.createOrLoadFromGeneric(realm, item);
                    incDecCell.setItem(itemDB);
                }
                realm.commitTransaction();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {}
        });
        itemAdapter = new ArrayAdapter<>(getActivity(), android.R.layout.simple_spinner_dropdown_item, items);
        sprItems.setAdapter(itemAdapter);
        List<ServerDB> servers = realm.where(ServerDB.class).findAll();
        items.clear();

        if(item != null){
            items.add(item);
            itemAdapter.add(item);
            itemAdapter.notifyDataSetChanged();
        }

        if(incDecCell.getItem() != null) {
            items.add(incDecCell.getItem().toGeneric());
        }
        for(final ServerDB serverDB : servers) {
            final OHServer server = serverDB.toGeneric();
            IServerHandler serverHandler = connectionFactory.createServerHandler(server, getContext());
            serverHandler.requestItemsRx()
                    .map(this::filterItems)
                    .compose(bindToLifecycle())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(newItems -> {
                        this.items.addAll(newItems);
                        itemAdapter.notifyDataSetChanged();
                    });
        }

        updateIconImage();
        btnSetIcon.setOnClickListener(v -> {
            Intent intent = new Intent(getActivity(), IconPickerActivity.class);
            startActivityForResult(intent, REQUEST_ICON);
        });

        return rootView;
    }

    private void updateIconImage(){
        btnSetIcon.setImageDrawable(Util.getIconDrawable(getActivity(), incDecCell.getIcon()));
    }

    private List<OHItem> filterItems(List<OHItem> items){

        List<OHItem> tempItems = new ArrayList<>();
        for(OHItem item : items){
            if(Constants.SUPPORT_INC_DEC.contains(item.getType())){
                tempItems.add(item);
            }
        }
        items.clear();
        items.addAll(tempItems);

        return items;
    }

    @Override
    public void onPause() {
        super.onPause();

        realm.beginTransaction();
        try {
            incDecCell.setMax(Integer.parseInt(txtMax.getText().toString()));
        }catch (NumberFormatException e) {
            incDecCell.setMax(100);
        }
        try {
            incDecCell.setMin(Integer.parseInt(txtMin.getText().toString()));
        }catch (NumberFormatException e) {
            incDecCell.setMin(0);
        }
        try {
            incDecCell.setValue(Integer.parseInt(txtValue.getText().toString()));
        }catch (NumberFormatException e) {
            incDecCell.setValue(1);
        }
        realm.commitTransaction();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        realm.close();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if(requestCode == REQUEST_ICON &&
                resultCode == Activity.RESULT_OK &&
                data.hasExtra(IconPickerActivity.RESULT_ICON)){

            String iconName = data.getStringExtra(IconPickerActivity.RESULT_ICON);
            realm.beginTransaction();
            incDecCell.setIcon(iconName.equals("") ? null : iconName);
            realm.commitTransaction();
            updateIconImage();
        }
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/control/cells/config/CellIncDecConfigFragment.kt
package treehou.se.habit.ui.control.cells.config

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.trello.rxlifecycle2.components.support.RxFragment
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import io.realm.Realm
import kotlinx.android.synthetic.main.inc_dec_controller_action.*
import se.treehou.ng.ohcommunicator.connector.models.OHItem
import treehou.se.habit.R
import treehou.se.habit.core.db.model.ItemDB
import treehou.se.habit.core.db.model.ServerDB
import treehou.se.habit.core.db.model.controller.CellDB
import treehou.se.habit.core.db.model.controller.IncDecCellDB
import treehou.se.habit.ui.util.IconPickerActivity
import treehou.se.habit.util.ConnectionFactory
import treehou.se.habit.util.Constants
import treehou.se.habit.util.Util
import treehou.se.habit.util.logging.Logger
import java.util.*
import javax.inject.Inject

class CellIncDecConfigFragment : RxFragment() {

    @Inject lateinit var connectionFactory: ConnectionFactory
    @Inject lateinit var logger: Logger
    lateinit var realm: Realm

    private var itemAdapter: ArrayAdapter<OHItem>? = null
    private val items = ArrayList<OHItem>()

    private var incDecCell: IncDecCellDB? = null
    private var cell: CellDB? = null
    private var item: OHItem? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        Util.getApplicationComponent(this).inject(this)
        realm = Realm.getDefaultInstance()

        if (arguments != null) {
            val id = arguments!!.getLong(ARG_CELL_ID)
            cell = CellDB.load(realm, id)
            incDecCell = cell!!.getCellIncDec()

            if (incDecCell == null) {
                realm.executeTransaction { realm ->
                    incDecCell = IncDecCellDB()
                    incDecCell = realm.copyToRealm(incDecCell!!)
                    cell!!.setCellIncDec(incDecCell!!)
                    realm.copyToRealmOrUpdate(cell!!)
                }
            }

            val itemDB = incDecCell!!.item
            if (itemDB != null) {
                item = itemDB.toGeneric()
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        return inflater.inflate(R.layout.inc_dec_controller_action, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        maxText.setText("" + incDecCell!!.max)
        minText.setText("" + incDecCell!!.min)
        valueText.setText("" + incDecCell!!.value)

        itemsSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
                realm.beginTransaction()
                val item = items[position]
                if (item != null) {
                    val itemDB = ItemDB.createOrLoadFromGeneric(realm, item)
                    incDecCell!!.item = itemDB
                }
                realm.commitTransaction()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {}
        }
        itemAdapter = ArrayAdapter(activity!!, android.R.layout.simple_spinner_dropdown_item, items)
        itemsSpinner.adapter = itemAdapter
        val servers = realm.where(ServerDB::class.java).findAll()
        items.clear()

        val item = item
        if (item != null) {
            items.add(item)
            itemAdapter!!.add(item)
            itemAdapter!!.notifyDataSetChanged()
        }

        if (incDecCell!!.item != null) {
            items.add(incDecCell!!.item!!.toGeneric())
        }
        for (serverDB in servers) {
            val server = serverDB.toGeneric()
            val serverHandler = connectionFactory.createServerHandler(server, context)
            serverHandler.requestItemsRx()
                    .map<List<OHItem>>({ this.filterItems(it) })
                    .compose(bindToLifecycle())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ newItems ->
                        this.items.addAll(newItems)
                        itemAdapter!!.notifyDataSetChanged()
                    }, { logger.e(TAG, "Failed to load items", it) })
        }

        updateIconImage()
        setIconButton.setOnClickListener {
            val intent = Intent(activity, IconPickerActivity::class.java)
            startActivityForResult(intent, REQUEST_ICON)
        }
    }

    private fun updateIconImage() {
        setIconButton.setImageDrawable(Util.getIconDrawable(activity, incDecCell!!.icon))
    }

    private fun filterItems(items: MutableList<OHItem>): List<OHItem> {

        val tempItems = ArrayList<OHItem>()
        for (item in items) {
            if (Constants.SUPPORT_INC_DEC.contains(item.type)) {
                tempItems.add(item)
            }
        }
        items.clear()
        items.addAll(tempItems)

        return items
    }

    override fun onPause() {
        super.onPause()

        realm.beginTransaction()
        try {
            incDecCell!!.max = Integer.parseInt(maxText.text.toString())
        } catch (e: NumberFormatException) {
            incDecCell!!.max = 100
        }

        try {
            incDecCell!!.min = Integer.parseInt(minText.text.toString())
        } catch (e: NumberFormatException) {
            incDecCell!!.min = 0
        }

        try {
            incDecCell!!.value = Integer.parseInt(valueText.text.toString())
        } catch (e: NumberFormatException) {
            incDecCell!!.value = 1
        }

        realm.commitTransaction()
    }

    override fun onDestroy() {
        super.onDestroy()

        realm.close()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_ICON &&
                resultCode == Activity.RESULT_OK &&
                data!!.hasExtra(IconPickerActivity.RESULT_ICON)) {

            val iconName = data.getStringExtra(IconPickerActivity.RESULT_ICON)
            realm.beginTransaction()
            incDecCell!!.icon = if (iconName == "") null else iconName
            realm.commitTransaction()
            updateIconImage()
        }
    }

    companion object {

        private val TAG = "CellIncDecConfigFragment"

        private val ARG_CELL_ID = "ARG_CELL_ID"
        private val REQUEST_ICON = 183

        fun newInstance(cell: CellDB): CellIncDecConfigFragment {
            val fragment = CellIncDecConfigFragment()
            val args = Bundle()
            args.putLong(ARG_CELL_ID, cell.id)
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
